appinventor/blocklyeditor/src/workspace_svg.js

```diff
@@ -608,20 +608,53 @@ Blockly.WorkspaceSvg.prototype.customContextMenu = function(menuOptions) {
   /**
    * Function that returns a name to be used to sort blocks.
    * The general comparator is the block.category attribute.
-   * In the case of 'Components' the comparator is the instanceName of the component if it exists
-   * (it does not exist for generic components).
    * In the case of Procedures the comparator is the NAME(for definitions) or PROCNAME (for calls)
+   * In the case of 'Components' the comparator is the type name, instance name, then event name
    * @param {!Blockly.Block} block the block that will be compared in the sortByCategory function
    * @returns {string} text to be used in the comparison
    */
   function comparisonName(block){
-    if (block.category === 'Component' && block.instanceName)
-      return block.instanceName;
-    if (block.category === 'Procedures')
-      return (block.getFieldValue('NAME') || block.getFieldValue('PROCNAME'));
-    return block.category;
+    // Add trailing numbers to represent their sequence
+    if (block.category == 'Variables'){
+      return ('1,' + block.type);
+    }
+    if (block.category === 'Procedures'){
+      return ('2,'+ (block.getFieldValue('NAME') || block.getFieldValue('PROCNAME')));
+    }
+    if (block.category == 'Component'){
+      var component = block.type + ',' + block.instanceName + ',' + block.eventName;
+      return ('3,' + component);
+    }
+    // Floating blocks that are not Component
+    return ('4,' + block.type);
   }
 
+  /**
+   * Function used to compare two strings with text and numbers
+   * @param {string} a first string to be compared
+   * @param {string} b second string to be compared
+   * @returns {number} returns 0 if the strings are equal, and -1 or 1 if they are not
+   */
+  function compareStrTextNum(strA,strB){
+    // Use Regular Expression to match text and numbers
+    var regexStrA = strA.match(/(.*?)([0-9]+)$/gi);
+    var regexStrB = strB.match(/(.*?)([0-9]+)$/gi);
+
+    // There are numbers in the strings, compare numbers
+    if (regexStrA != null && regexStrB != null){
+      numStrA = strA.match(/([0-9]+)$/gi);
+      numStrB = strB.match(/([0-9]+)$/gi);
+      numA = parseInt(numStrA);
+      numB = parseInt(numStrB);
+      if (numA > numB) return +1;
+      else if (numA < numB) return -1;
+      else return 0;
+    }else{
+      return strA.localeCompare(strB);
```

Comment:
Wherever you use localeCompare, you should use the 3-parameter variant: `first.localeCompare(second, undefined, {number: true})`. This eliminates the need to use the regex above.

Reply:
Correction: the key in the options dictionary should be `numeric`, not `number`